File: Problems/Algorithms/5. Longest Palindromic Substring/LongestPalindrome.java
class Solution {
    public String longestPalindrome(String s) {
        int length = 0;
        int n = s.length();
        
        int start = 0;
        int end = n-1;
        for (int i = 0; i < n; i++) {
            int left = i;
            int right = i;
            while (left >= 0 && right < n && s.charAt(left) == s.charAt(right)) {
                if (right - left + 1 > length) {
                    start = left;
                    end = right;
                    length = right - left + 1;
                }
                left--;
                right++;
            }
            
            left = i;
            right = i + 1;
            while (left >= 0 && right < n && s.charAt(left) == s.charAt(right)) {
                if (right - left + 1 > length) {
                    start = left;
                    end = right;
                    length = right - left + 1;
                }
                left--;
                right++;
            }
        }
            
        return s.substring(start, end + 1);
    }
}

File: Problems/Algorithms/5. Longest Palindromic Substring/longest_palindrome.rs
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        for size in (1..=s.len()).rev() {
            match s.as_bytes()
                .windows(size)
                .find(|substr| Self::is_palindrome(substr)) {
                    Some(palindrome) => return String::from_utf8(palindrome.to_vec()).unwrap(),
                    None => continue,
            }
        }
        
        String::from("")
    }
    
    pub fn is_palindrome(s: &[u8]) -> bool {
        s.iter().zip(s.iter().rev()).all(|(left, right)| left == right)
    }
}

